File: test/python/steps/draft_steps.py
# -*- coding: utf-8 -*-
"""
Draft save and resume step definitions.

These steps test the draft functionality that allows applicants to save
partial applications and resume them later after logout/login.
"""
from __future__ import absolute_import
from behave import given, when, then


# ============================================================================
# DRAFT SAVE AND RESUME STEPS
# ============================================================================

@when('I fill in the TTC application form partially with:')
@when('I fill in the TTC application form partially with')
def step_fill_partial_form(context, doc=None):
    """Store partial form data in context for later save."""
    # Handle both table parameter and non-table parameter calls
    if doc is None:
        doc = context.table if hasattr(context, 'table') else None

    # Parse the table data into a dictionary
    context.partial_form_data = {}
    if doc:
        for row in doc.rows:
            context.partial_form_data[row['field']] = row['value']

    # Initialize draft storage for user if not exists
    if not hasattr(context, 'drafts'):
        context.drafts = {}

    # Store in drafts under 'ttc_application' key
    context.drafts['ttc_application'] = {
        'form_type': 'ttc_application',
        'status': 'draft',
        'data': context.partial_form_data.copy()
    }


@when('I save the application as draft')
def step_save_draft(context):
    """Simulate saving draft to persistent storage."""
    # Verify partial data exists
    assert hasattr(context, 'drafts'), 'No draft data to save'
    assert 'ttc_application' in context.drafts, 'No TTC application draft to save'

    # Mark draft as saved
    context.drafts['ttc_application']['saved'] = True
    context.drafts['ttc_application']['saved_at'] = 'now'


@then('I should see my draft data persisted')
def step_draft_persisted(context):
    """Verify draft data was saved and can be retrieved."""
    assert hasattr(context, 'drafts'), 'No drafts found'
    assert 'ttc_application' in context.drafts, 'No TTC application draft found'
    assert context.drafts['ttc_application'].get('saved') is True, 'Draft was not saved'

    # Verify original data is intact
    assert 'data' in context.drafts['ttc_application'], 'Draft has no data'
    draft_data = context.drafts['ttc_application']['data']

    # Check for expected fields from scenario
    assert 'i_fname' in draft_data or 'i_lname' in draft_data or 'i_email' in draft_data, \
        'Draft data missing expected fields'


@when('I complete the remaining required fields and submit')
def step_complete_and_submit(context):
    """Complete the form with remaining fields and submit."""
    # Get existing draft data
    if not hasattr(context, 'drafts') or 'ttc_application' not in context.drafts:
        context.drafts = {'ttc_application': {'data': {}}}

    # Add remaining required fields for a complete submission
    required_fields = {
        'i_address1': '123 Main St',
        'i_city': 'Springfield',
        'i_state': 'IL',
        'i_zip': '62701',
        'i_phone': '555-123-4567',
        'i_gender': 'prefer_not_to_say',
    }

    # Merge with existing draft data
    context.drafts['ttc_application']['data'].update(required_fields)

    # Mark as submitted
    context.drafts['ttc_application']['status'] = 'submitted'
    context.drafts['ttc_application']['submitted_at'] = 'now'

    # Set last_submission for compatibility with existing assertion steps
    context.last_submission = {
        'form_type': 'ttc_application',
        'status': 'submitted',
        'data': context.drafts['ttc_application']['data'].copy()
    }


@when('I open the TTC application form')
def step_open_ttc_form(context):
    """Open the TTC application form (generic, not country-specific)."""
    context.current_form = 'ttc_application'


@when('I save a partial TTC application as draft')
def step_save_partial_ttc_draft(context):
    """Save a partial TTC application with minimal data."""
    if not hasattr(context, 'drafts'):
        context.drafts = {}

    context.drafts['ttc_application'] = {
        'form_type': 'ttc_application',
        'status': 'draft',
        'saved': True,
        'data': {
            'i_fname': 'Test',
            'i_lname': 'Applicant',
            'i_email': 'test.applicant@example.com',
        }
    }


@when('I save a partial evaluator profile as draft')
def step_save_partial_evaluator_draft(context):
    """Save a partial evaluator profile with minimal data."""
    if not hasattr(context, 'drafts'):
        context.drafts = {}

    context.drafts['evaluator_profile'] = {
        'form_type': 'evaluator_profile',
        'status': 'draft',
        'saved': True,
        'data': {
            'ev_fname': 'Test',
            'ev_lname': 'Evaluator',
            'ev_email': 'test.evaluator@example.com',
            'ev_organization': 'Test Organization',
        }
    }




@then('I should see the TTC application draft data')
def step_see_ttc_draft_data(context):
    """Verify TTC application draft data is visible."""
    assert hasattr(context, 'drafts'), 'No drafts found'
    assert 'ttc_application' in context.drafts, 'No TTC application draft found'

    draft = context.drafts['ttc_application']
    assert draft.get('status') == 'draft', 'Expected draft status'
    assert 'data' in draft, 'Draft has no data'

    # Verify expected fields exist
    data = draft['data']
    assert 'i_fname' in data or 'i_lname' in data, 'Missing expected draft fields'


@when('I navigate to the evaluator profile form')
def step_navigate_to_evaluator_form(context):
    """Set current form context to evaluator profile."""
    context.current_form = 'evaluator_profile'


@then('I should see the evaluator profile draft data')
def step_see_evaluator_draft_data(context):
    """Verify evaluator profile draft data is visible."""
    assert hasattr(context, 'drafts'), 'No drafts found'
    assert 'evaluator_profile' in context.drafts, 'No evaluator profile draft found'

    draft = context.drafts['evaluator_profile']
    assert draft.get('status') == 'draft', 'Expected draft status'
    assert 'data' in draft, 'Draft has no data'

    # Verify expected fields exist
    data = draft['data']
    assert 'ev_fname' in data or 'ev_lname' in data, 'Missing expected draft fields'

@when('I test fill without colon')
def step_test_fill_no_colon(context):
    """Test step without colon."""
    pass

File: test/typescript/steps/draft_steps.ts
/**
 * Draft save and resume step definitions.
 *
 * These steps test the draft functionality that allows applicants to save
 * partial applications and resume them later after logout/login.
 */

import { DataTable, When, Then } from '@cucumber/cucumber';
import assert from 'assert';

// ============================================================================
// TYPE DEFINITIONS
// ============================================================================

interface DraftData {
  form_type: string;
  status: 'draft' | 'submitted';
  saved?: boolean;
  saved_at?: string;
  submitted_at?: string;
  data: Record<string, string>;
}

// ============================================================================
// DRAFT CONTEXT
// ============================================================================

const draftContext: {
  drafts: Record<string, DraftData>;
  currentForm?: string;
  partialFormData?: Record<string, string>;
} = {
  drafts: {},
  currentForm: undefined,
  partialFormData: undefined,
};

// ============================================================================
// DRAFT SAVE AND RESUME STEPS
// ============================================================================

When('I fill in the TTC application form partially with:', (dataTable: DataTable) => {
  const rows = dataTable.hashes();
  const partialData: Record<string, string> = {};

  for (const row of rows) {
    partialData[row.field] = row.value;
  }

  // Store in context
  draftContext.partialFormData = partialData;

  // Initialize draft storage for user if not exists
  if (!draftContext.drafts['ttc_application']) {
    draftContext.drafts['ttc_application'] = {
      form_type: 'ttc_application',
      status: 'draft',
      data: {},
    };
  }

  // Store partial data in drafts
  draftContext.drafts['ttc_application'].data = { ...partialData };
});

When('I save the application as draft', () => {
  assert(draftContext.drafts['ttc_application'], 'No TTC application draft to save');

  const draft = draftContext.drafts['ttc_application'];
  draft.saved = true;
  draft.saved_at = new Date().toISOString();
});

Then('I should see my draft data persisted', () => {
  assert(draftContext.drafts['ttc_application'], 'No TTC application draft found');
  const draft = draftContext.drafts['ttc_application'];

  assert.strictEqual(draft.saved, true, 'Draft was not saved');
  assert(draft.data, 'Draft has no data');

  // Verify at least one expected field exists
  const hasData = Object.keys(draft.data).some(key =>
    ['i_fname', 'i_lname', 'i_email'].includes(key)
  );
  assert(hasData, 'Draft data missing expected fields');
});

When('I complete the remaining required fields and submit', () => {
  if (!draftContext.drafts['ttc_application']) {
    draftContext.drafts['ttc_application'] = {
      form_type: 'ttc_application',
      status: 'draft',
      data: {},
    };
  }

  const requiredFields: Record<string, string> = {
    i_address1: '123 Main St',
    i_city: 'Springfield',
    i_state: 'IL',
    i_zip: '62701',
    i_phone: '555-123-4567',
    i_gender: 'prefer_not_to_say',
  };

  // Merge with existing data
  Object.assign(draftContext.drafts['ttc_application'].data, requiredFields);

  // Mark as submitted
  draftContext.drafts['ttc_application'].status = 'submitted';
  draftContext.drafts['ttc_application'].submitted_at = new Date().toISOString();
});

When('I save a partial TTC application as draft', () => {
  draftContext.drafts['ttc_application'] = {
    form_type: 'ttc_application',
    status: 'draft',
    saved: true,
    data: {
      i_fname: 'Test',
      i_lname: 'Applicant',
      i_email: 'test.applicant@example.com',
    },
  };
});

When('I save a partial evaluator profile as draft', () => {
  draftContext.drafts['evaluator_profile'] = {
    form_type: 'evaluator_profile',
    status: 'draft',
    saved: true,
    data: {
      ev_fname: 'Test',
      ev_lname: 'Evaluator',
      ev_email: 'test.evaluator@example.com',
      ev_organization: 'Test Organization',
    },
  };
});

When('I navigate to the TTC application form', () => {
  draftContext.currentForm = 'ttc_application';
});

When('I open the TTC application form', () => {
  draftContext.currentForm = 'ttc_application';
});

Then('I should see the TTC application draft data', () => {
  assert(draftContext.drafts['ttc_application'], 'No TTC application draft found');
  const draft = draftContext.drafts['ttc_application'];

  assert.strictEqual(draft.status, 'draft', 'Expected draft status');
  assert(draft.data, 'Draft has no data');

  const hasExpectedField = ['i_fname', 'i_lname', 'i_email'].some(key => key in draft.data);
  assert(hasExpectedField, 'Missing expected draft fields');
});

When('I navigate to the evaluator profile form', () => {
  draftContext.currentForm = 'evaluator_profile';
});

Then('I should see the evaluator profile draft data', () => {
  assert(draftContext.drafts['evaluator_profile'], 'No evaluator profile draft found');
  const draft = draftContext.drafts['evaluator_profile'];

  assert.strictEqual(draft.status, 'draft', 'Expected draft status');
  assert(draft.data, 'Draft has no data');

  const hasExpectedField = ['ev_fname', 'ev_lname', 'ev_email'].some(key => key in draft.data);
  assert(hasExpectedField, 'Missing expected draft fields');
});

// Export the draft context for use in other step files if needed
export { draftContext, DraftData };
